File: frontend/src/utils/crypto.ts
export function generateKey(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(32));
}

export async function encrypt(
  key: Uint8Array,
  plaintext: string
): Promise<string> {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const encoded = new TextEncoder().encode(plaintext);

  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    key.buffer as ArrayBuffer,
    { name: "AES-GCM" },
    false,
    ["encrypt"]
  );

  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: iv.buffer as ArrayBuffer },
    cryptoKey,
    encoded
  );

  const combined = new Uint8Array(iv.length + ciphertext.byteLength);
  combined.set(iv, 0);
  combined.set(new Uint8Array(ciphertext), iv.length);

  return btoa(String.fromCharCode(...combined))
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

export async function decrypt(
  key: Uint8Array,
  encryptedB64: string
): Promise<string> {
  const padded = encryptedB64.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(padded);
  const bytes = Uint8Array.from(binary, (c) => c.charCodeAt(0));

  const iv = bytes.slice(0, 12);
  const ciphertext = bytes.slice(12);

  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    key.buffer as ArrayBuffer,
    { name: "AES-GCM" },
    false,
    ["decrypt"]
  );

  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: iv.buffer as ArrayBuffer },
    cryptoKey,
    ciphertext
  );

  return new TextDecoder().decode(decrypted);
}

export function encodeKey(key: Uint8Array): string {
  return btoa(String.fromCharCode(...key))
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

export function decodeKey(keyB64: string): Uint8Array {
  const padded = keyB64.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(padded);
  return Uint8Array.from(binary, (c) => c.charCodeAt(0));
}

export async function hashPassword(password: string): Promise<{ hash: string; salt: string }> {
  const encoder = new TextEncoder();
  const salt = crypto.getRandomValues(new Uint8Array(16));

  const keyMaterial = await crypto.subtle.importKey(
    "raw",
    encoder.encode(password),
    { name: "PBKDF2" },
    false,
    ["deriveBits"]
  );

  const hashBuffer = await crypto.subtle.deriveBits(
    {
      name: "PBKDF2",
      salt: salt,
      iterations: 100000,
      hash: "SHA-256",
    },
    keyMaterial,
    256
  );

  const hashArray = new Uint8Array(hashBuffer);
  const hashHex = Array.from(hashArray).map(b => b.toString(16).padStart(2, '0')).join('');
  const saltHex = Array.from(salt).map(b => b.toString(16).padStart(2, '0')).join('');

  return { hash: hashHex, salt: saltHex };
}

File: cli/poof/crypto.py
import os
import base64
from nacl.secret import SecretBox
from nacl.utils import random as nacl_random


def generate_key() -> bytes:
    """Generate a random 256-bit key."""
    return nacl_random(SecretBox.KEY_SIZE)


def encrypt(key: bytes, plaintext: str) -> str:
    """Encrypt plaintext. Returns base64-encoded ciphertext."""
    box = SecretBox(key)
    nonce = nacl_random(box.NONCE_SIZE)
    ciphertext = box.encrypt(plaintext.encode(), nonce)
    return base64.urlsafe_b64encode(ciphertext).decode()


def decrypt(key: bytes, encrypted_b64: str) -> str:
    """Decrypt base64-encoded ciphertext."""
    box = SecretBox(key)
    ciphertext = base64.urlsafe_b64decode(encrypted_b64)
    plaintext = box.decrypt(ciphertext)
    return plaintext.decode()


def encode_key(key: bytes) -> str:
    """Encode key to base64url string for URL fragment."""
    return base64.urlsafe_b64encode(key).decode().rstrip("=")


def decode_key(key_b64: str) -> bytes:
    """Decode base64url key from URL fragment."""
    padded = key_b64 + "=" * (4 - len(key_b64) % 4)
    return base64.urlsafe_b64decode(padded)
